Expose keepOpenOnPeerClosed for outgoing TCP streams (#22163)

raboof opened the pull request:
I'm not too familiar with this part of the codebase but it seems to make sense
to expose this option. Not sure what the most sensible default is: accidentally
enabling it might cause missing data when the server is eager in closing the
connections, accidentally disabling it might cause the outgoing side of the
connection to 'leak'.

raboof committed:
Expose keepOpenOnPeerClosed for outgoing TCP streams (#22163)

I'm not too familiar with this part of the codebase but it seems to make sense
to expose this option. Not sure what the most sensible default is: accidentally
enabling it might cause missing data when the server is eager in closing the
connections, accidentally disabling it might cause the outgoing side of the
connection to 'leak'.

raboof commented:
Refs #22163

jrudolph reviewed (approved):
LGTM.

Not sure how to fix those bin compat APIs. I think we should avoid default arguments in APIs in general because it's impossible to evolve those APIs otherwise.

raboof committed:
Binary compatibility

Introduce a new name for opening a connection, switch
the default to automatically closing the client side when
the server side terminates.

raboof committed:
Missing import in documentation code

raboof committed:
More binary compatibility updates

raboof committed:
Correctly register MiMa filter

raboof committed:
Moved MiMa filter to the right place

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Tcp.scala:
```diff
@@ -144,6 +144,17 @@ class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
       .map(new IncomingConnection(_))
       .mapMaterializedValue(_.map(new ServerBinding(_))(ec).toJava))
 
+  @Deprecated
```
add description and version as in other places

patriknw replied:
ah, this is the java annotation, hm I think we use the scala annotation for such things (I hope that will show up as deprecated when used from java)

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Tcp.scala:
```diff
@@ -159,23 +170,32 @@ class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
+    halfClose:            Boolean,
+    connectTimeout:       Duration,
+    idleTimeout:          Duration,
+    keepOpenOnPeerClosed: Boolean): Flow[ByteString, ByteString, CompletionStage[OutgoingConnection]] =
```
if we are anyway changing this it might be worth creating a settings class for these options. Easier to evolve. options, halfClose, connectTimeout, idleTimeout, keepOpenOnPeerClosed could all go into that

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Tcp.scala:
```diff
@@ -159,23 +170,32 @@ class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
    */
-  def outgoingConnection(host: String, port: Int): Flow[ByteString, ByteString, CompletionStage[OutgoingConnection]] =
-    Flow.fromGraph(delegate.outgoingConnection(new InetSocketAddress(host, port))
+  def connect(host: String, port: Int): Flow[ByteString, ByteString, CompletionStage[OutgoingConnection]] =
```
naming is hard, since it is not really performing the connect at this point it might be misleading.

patriknw commented:
what is the status here? other opinions about the naming?

patriknw commented:
@raboof wow, 2 years ago, maybe you can review this yourself to get a second approval ;-)

jrudolph commented on akka-stream-tests/src/test/scala/akka/stream/io/TcpSpec.scala:
```diff
@@ -45,11 +45,15 @@ class TcpSpec extends StreamSpec("""
 
       val tcpReadProbe = new TcpReadProbe()
       val tcpWriteProbe = new TcpWriteProbe()
+<<<<<<< HEAD
```
Oops, unresolved conflicts.

raboof commented:
If anyone finds this issue and would like to see this feature feel free to reopen, I'd like to get back to it but it's not high on the list right now.
